### ISSUE TITLE
Code Review Policy in DevTools

### ISSUE THREAD

**Issue by gregtatum** [devtools-html/rfcs#41]
This RFC is a follow-up with #37.

It would be nice to re-visit our code review practices, and update our [internal documentation](http://docs.firefox-dev.tools/contributing/code-reviews.html). Right now we have a checklist to specific issues to look for, which can be helpful for new contributors, but we don't address _why_ we do code reviews. I think we are missing a higher level policy document. 

I personally think this is really important discussion to have as a group, because every day we have the code that we spent our time, passion, and energy on criticized by our peers. This is really important for writing high quality code, but it can be a difficult part of our jobs as we are personally invested in our little diffs of code. In addition, as a really distributed team, it's easy to lack empathy with each other when all you are presented with is a list of issues to fix from a glowing screen.

@jryans shared "[How to Do Code Reviews Like a Human](https://mtlynch.io/human-code-reviews-1/)", and I also did some of my own research for different recommendations. I'm going to try and summarize some specific recommendations with headlines, and include a snippet of support from various articles. Feel free to skim the quotes and focus on the headlines, although I would recommend reading the full articles.

#### Discussion

In your response please copy and paste this format to help structure this discussion, so that hopefully we can come to some type of consensus.

```
## What is the answer to: "Why do we do code reviews?"

## Empathetic and inclusive language

## Scope of requested changes

## Prefer small PRs, and request to split up large reviews

## Be explicit about required and optional changes

## Are there any missing top level items?

## Additional comments
```

#### Relevant reading

 * [How to Do Code Reviews Like a Human](https://mtlynch.io/human-code-reviews-1/)
 * [The Pitfalls of Code Review (And How To Fix Them)](https://www.brandonsavage.net/the-pitfalls-of-code-review-and-how-to-fix-them/)
 * Does anyone else have more recommended reading?

----
 
## What is the answer to: "Why do we do code reviews?"

I've heard surprisingly different things from people on this, and I think it would be beneficial to hear from DevToolers on why we do code reviews. I'll add mine as a follow-up comment.

## Empathetic and inclusive language

I think we should have a specific recommendation for using empathetic and inclusive language in our code reviews.

#### [Frame feedback as requests, not commands](https://mtlynch.io/human-code-reviews-1/#frame-feedback-as-requests-not-commands)

> Code reviews require more tact and care than usual communication because there’s a high risk of derailing the discussion into a personal argument. You would expect reviewers to dial up their politeness in reviews, but bizarrely I’ve found them to go the opposite direction. Most people never say to a co-worker, “Hand me that stapler, then fetch me a soda.” But I’ve seen numerous reviewers frame feedback with similarly pushy commands, such as, “Move this class to a separate file.”
>
> Err on the side of being annoyingly gentle in your feedback. Frame your notes as requests or suggestions, not commands.
>
> [(Read more...)](https://mtlynch.io/human-code-reviews-1/#frame-feedback-as-requests-not-commands)

#### [Never say “you”](https://mtlynch.io/human-code-reviews-1/#never-say-you)

> This one is going to sound weird, but hear me out: never use the word “you” in a code review.
>
> The decisions you reach in a review should be based on what makes the code better rather than who came up with the idea. Your teammate put significant effort into their changelist and is likely proud of the work they did. Their natural reaction to hearing criticism of their work is to feel defensive and protective.
>
> Word your feedback in a way that minimizes the risk of raising your teammate’s defenses. Be clear that you’re critiquing the code, not the coder. When an author sees “you” in a comment, it brings their focus away from the code and back to themselves. This increases the risk that they’ll take your criticism personally.
>
> [(Read more...)](https://mtlynch.io/human-code-reviews-1/#never-say-you)

Although apparently some folks found the indirect communication style difficult:

> After I published the first half of this article, several readers took issue with the communication style I recommended. Some found it patronizing. Others worried that it was too indirect and risked miscommunication.
>
> That feedback is reasonable and expected. One person may find a terse review comment to be brusque or rude. Another may judge the same comment as concise and efficient.
>
> In reviewing code, you make many choices: what to focus on, how to frame feedback, when to approve. It’s not important that you choose my options. Just recognize that there are options.

## Scope of requested changes

Another theme is to not block a review to ask someone to fix unrelated changes to their patch. These issues should be filed separately.

#### [Respect the scope of the review](https://mtlynch.io/human-code-reviews-2/#respect-the-scope-of-the-review)

> There’s an anti-pattern I see frequently where the reviewer identifies something near code in the changelist and asks the author to fix it. Once the author complies, the reviewer usually realizes that the code is better but inconsistent, so it needs a few more minor changes. And then a few more. And on and on until a narrowly-scoped changelist has expanded to include lots of unrelated churn.
>
> The rule of thumb is: if the changelist doesn’t touch the line, it’s out of scope.
>
> [(Read more..)](https://mtlynch.io/human-code-reviews-2/#respect-the-scope-of-the-review)

## Prefer small PRs, and request to split up large reviews

Smaller PRs are easier to review and get consensus over.

#### [Look for opportunities to split up large reviews](https://mtlynch.io/human-code-reviews-2/#look-for-opportunities-to-split-up-large-reviews)

> If you receive a changelist that’s more than ~400 lines of code, encourage the author to split it into smaller pieces. Push back proportionally harder the more they go over this limit. I personally refuse to review any changelists that exceed 1,000 lines.
>
> When the code quality is low, emphatically request a split. The difficulty of reviewing bad code grows exponentially with size. You’re much better off auditing a couple of sloppy 300-line changelists than a single 600-line abomination.

## Be explicit about required and optional changes

When there are lots of changes to address, it should be clear which are optional and which are mandatory. When reviewing code it's easy to comment across large parts of the codebase with suggestions and observations. It should be clear when it's mandatory for your review, and when it's something that's optional.

I think this has some follow up questions that might be helpful to answer.

 * How do we actually resolve conflict of when the author and reviewer disagree?
 * When is it OK to merge something when you disagree with what a reviewer says?
 * How much do we trust that the author of the code will do the right thing? Do we have to follow-up with every change before approval?
 * Is it OK to let small code quality issues slip in?
 * Is it OK to fail forward, and fix things in follow-ups?

#### [Grant approval when remaining fixes are trivial](https://mtlynch.io/human-code-reviews-2/#grant-approval-when-remaining-fixes-are-trivial)

> Some reviewers have the misconception that they should withhold approval until they witness fixes for every last note. This adds needless code review rounds, wasting time for both author and reviewer.
>
> Grant approval when any of the following are true:
>
>    You have no more notes.
>    Your remaining notes are for trivial issues.
>        E.g., renaming a variable, fixing a typo
>    Your remaining notes are optional suggestions.
>        Explicitly mark these as optional so that your teammate doesn’t assume your approval is contingent on them.
>
> I’ve seen reviewers withhold approval because the author missed a period at the end of a code comment. Please don’t do this. It signals to the author that you think they’re incapable of adding simple punctuation unless supervised.
>
> There is some danger in granting approval when there are still outstanding notes. I estimate that ~5% of the time, the author either misinterprets a final round note or misses it completely. To mitigate this, I simply check the author’s post-approval changes. In the rare case of miscommunication, I either follow up with the author or create my own changelist with a fix. Adding a small amount of work to the 5% case is better than adding unnecessary effort and delay to other 95%.

#### [Three-tier comment system]((https://www.brandonsavage.net/the-pitfalls-of-code-review-and-how-to-fix-them/))

> I use a three-tier system for code reviews. Comments that begin with “You might…” indicate that I am making a suggestion, based on my experience. The developer can choose to ignore my suggestion or not, based on their experience. “You should…” comments indicate something I feel strongly about. The developer is strongly encouraged to consider the suggestion, but again, I leave this to their discretion.
>
> Comments beginning with “You must…” are not optional. But these kinds of comments are reserved for security vulnerabilities, coding standard violations, obvious bugs, or other serious code problems. These are the only blocking code review comments, meaning that all other code review comments can be ignored, but these must be dealt with to obtain my approval for the patch. Everything else gets a thumbs-up.

> [(Read more...)](https://www.brandonsavage.net/the-pitfalls-of-code-review-and-how-to-fix-them/)

#### [Handle stalemates proactively](https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively)

> The worst possible outcome of a code review is a stalemate: you refuse to sign off on the changelist without further changes, but the author refuses to make them.
>
> Here are some indicators that you’re headed for a stalemate:
>
>    The tone of the discussion is growing tense or hostile.
>    Your notes per review round are not trending downward.
>    You’re getting pushback on an unusually high number of your notes.
> [(Read more...)](https://mtlynch.io/human-code-reviews-2/#handle-stalemates-proactively)

**Comment by julienw** [comment]
Thanks for this write-up, and for the link to this article I already knew but it's always good to read it again as it contains very thoughful suggestions.

Here my 2 cents:

## What is the answer to: "Why do we do code reviews?"
### Quality
We do code reviews to improve the overall quality of the code we ship to users. By better quality, I mean:

* less bugs
* better performance
* better UX and design

### Sharing knowledge
We also do code reviews to share the knowledge among our team (or at least subteams):
an inexperienced developer will ask a review from an experienced developer, and learn from the experienced developer's knowledge.

But what I find even better, and that we often forget is:
an experienced developer will ask a review from an inexperienced developer so that this inexperienced developer can build some knowledge around this area of code. Likely the inexperienced developer will scrutinize the code more and may find issues the experienced developer haven't thought of.


## Empathetic and inclusive language

I agree with what you write above. This is especially important when dealing with contributors that aren't Mozilla employees, so we need to take extra care in this situation.

However I also believe that in a professionnal setting where we know each other quite well, this should be less of a problem. This is also part of the job of the reviewee to know that in a review, the code is reviewed, not him/herself. Yes, sometimes we can feel hurt by some reviews (I felt it myself of course) but then we can bite the bullet and calm down., because we know the reviewer wasn't harsh on purpose.
Also everybody needs to remember that most of our team talks and writes english as a second language.

I guess my comment is that, yes, we should all take care to keep a polite and professional tone, but also accept that this might not be the case sometimes, when we know each other.

## Scope of requested changes

Completely agree.

## Prefer small PRs, and request to split up large reviews

Completely agree.

Splitting can be either in separate PR (prefered) or in several commits in one PR (more difficult but makes sense if they're related).

## Be explicit about required and optional changes

Completely agree.
I know that I myself am sometimes too implicit. Explicit is better, especially when (as said above) english is the second language for most of our team.

About `Grant approval when remaining fixes are trivial`, this is again a case where it can be different whether the reviewee is a long-term contributor or a newcomer.

As for your list of questions:
###    How do we actually resolve conflict of when the author and reviewer disagree?

I always took the decision that the reviewer is always right. Of course we can discuss, but in the end, if there's a conflict, I always obey the reviewer.

###    When is it OK to merge something when you disagree with what a reviewer says?

I'd say yes for an optional thing, but see the previous answer. Unless the reviewer said something like "you decide" or "up to you", I'll fix all comments he made, even optional ones.

###    How much do we trust that the author of the code will do the right thing? Do we have to follow-up with every change before approval?

I still usually have a look after the merge, that the author actually fixed the last comments ;) (Unless it's @gregtatum whom I blindly trust, of course)

###    Is it OK to let small code quality issues slip in?

Define "small" ;)
This is technical debt. Technical debt is when you borrow some code quality so that you can move faster, but at one moment you'll have to give it back. So I think it's OK if you keep this in your mind. What's your current technical debt ? Is it worth keeping this issue to move faster, or is it better/faster to fix it now ?

###    Is it OK to fail forward, and fix things in follow-ups?

As a reviewer, I usually say "ok, but file a bug and put the bug number as a comment in this code". So that at least we don't forget follow-ups, and when reading this code we know where to look at. Especially when we write a workaround for some bug, it's useful to know for which bug so that we know later more easily that the workaround can be removed.

## Are there any missing top level items?

I can't think of anything.

## Additional comments

Never forget that we can also discuss about patches live, either on IRC or Slack, or even on video chat. Sometimes we can move faster this way especially when there's a misunderstanding.

**Comment by jryans** [comment]
## What is the answer to: "Why do we do code reviews?"

I believe we do code reviews to ensure code changes are of a high quality and to confirm that they make sense to someone else on the team. Code reviews are also an opportunity for the reviewer to learn about the area being changed, in case they haven't worked on it directly or haven't seen it in a while. The reviewer can also act as a backup in case there are follow up questions or issues that arise later on and the patch author is unavailable.

In a ideal world, stylistic nits are handling by linting tools following patterns that the team has previously agreed on, so human review shouldn't need to worry about stylistic issues, and can instead focus on substance of the change. If a stylistic issue does come up, we should strive to think about how to automate for future reviews, instead of mandating mechanical checks by reviewers.

## Empathetic and inclusive language

I agree we should aim to be more empathetic and inclusive in our comment style, and the points here about using "we" and making requests seem like good steps towards that.

## Scope of requested changes

Agreed, we shouldn't require unrelated changes. The reviewer can request a follow up issue to be filed or file such an issue themselves, whichever seems appropriate at the time.

## Prefer small PRs, and request to split up large reviews

Whenever reasonable, yes, we should keep reviews smaller. Throwing a massive review at someone is almost like cheating because it's quite unlikely the review will be done with the same level of attention as a reasonably sized patch. Using a stack of multiple well-contained commits (as opposed to one patch for all changes, no matter the size) seems to work well for a variety of changes.

See also the "microcommit" philosophy suggested by the [MozReview user guide](https://mozilla-version-control-tools.readthedocs.io/en/latest/mozreview/commits.html#prefer-more-smaller-commits-over-large-monolithic-commits):

> ## Prefer More, Smaller Commits Over Large, Monolithic Commits
>
> Performing code review is hard.
> 
> In general, code review becomes even harder the more content that is changed in a commit (because the reviewer has more to look at).
> 
> Reviewing a single diff that changes hundreds or even thousands of lines is a daunting and time-consuming task. Furthermore, the more that is changed, the higher the chances that there will be something that is wrong. The more iterations between reviewer and author there are, the longer it takes for changes to land and the longer others have to wait for progress to be made.
> 
> For these reasons and more, it is recommend to author more, smaller commits than fewer, larger commits. This practice is sometimes referred to as microcommits. In general, a commit should be as small as possible but no smaller. Here are some guidelines:
> 
> * If you need to perform some cleanup before a refactor, put the cleanup in its own commit separate from the refactor.
> * If you need to fix a typo, put that in its own commit.
> * If you need to make a wide-sweeping change such as adding an argument to a commonly-called function, update function declarations one at a time (1 per commit) or use 1 commit to introduce the new interface and another for implementing it.

I don't mean to suggest requiring many tiny commits, as that sounds too restrictive. Instead, all I mean is that it's a spectrum, and there are many situations where breaking up your work into several reasonable chunks will greatly aid reviewing.

## Be explicit about required and optional changes

For the headline issue here, yes, we should always be explicit about what's required vs. optional.

### How do we actually resolve conflict of when the author and reviewer disagree?

Where possible, we should try our best to reach some mutual understanding or compromise. Perhaps chatting about the issue on IRC / Slack / Vidyo may help. I haven't seen too many cases at Mozilla where the two parties vehemently disagree with no path of resolution.

### When is it OK to merge something when you disagree with what a reviewer says?

If it's an optional review comment, then it seems fine to ignore if you disagree, since after all it was stated as optional, but it's still nice to reply and explain your position. If it's a required review comment, then I think it's best to try to resolve the disagreement before merging. After all, both parties have put effort into the review process. It feels somewhat rude in my opinion to ignore required review comments without attempting further discussion.

### How much do we trust that the author of the code will do the right thing? Do we have to follow-up with every change before approval?

For new contributors, I generally want to verify they've made every change (and thus I am unlikely to give them "optional" comments). For familiar contributors (which includes staff as well), we should be able to trust them to do the right thing. Usually I only request another round from a familiar contributor if large design changes are needed.

### Is it OK to let small code quality issues slip in?

In some cases, yes, but ideally we're not doing so regularly. Code quality is a good value to strive for, but it's just another axis of product development like "time required to ship", "technical debt", etc. that the team should keep in mind and balance together as a group of humans working towards a common goal.

### Is it OK to fail forward, and fix things in follow-ups?

I am not totally sure what meaning of "fail" is implied here... If it means merging a patch that would cause some tests to fail and block others because CI is red, then I don't think we want to do that. If it means disabling some tests "temporarily", then it could be okay as long as the team understands the risk that fixing those tests can easily fall to the backlog and never be done.

## Are there any missing top level items?

I can't think of others. This list seems like a good set to me.

## Additional comments

Thanks for starting this discussion! 😁

**Comment by ochameau** [comment]
## What is the answer to: "Why do we do code reviews?"
To ensure that:
- one another person can understand the modification (the patch) and the final code (base code + the patch)
  To me, it is one of the most important point and I feel that we aren't perfect on this.
  Any code that isn't obvious should be at very least questioned during the review.
  Then must follow: code or logic simplification, adding comments, new docs, ...
- the modification implements what it originally aims to implement
  We should verify the logic implemented by the patch. To check that it does what it is meant to do. We shouldn't only run firefox with the patch to check if things still work.
- it doesn't break something
  Here I'm thinking about looking for edge cases, things that aren't tested. Like races, UX issues, exotic environment (launchpad, browser toolbox, remote debugging), ...
  This isn't about tests. @jryans suggests that "stylistic changes" should be handled by the developer, by running eslint himself. But I would like to extend that to tests. Try is here to verify the patches. It is up to the developer to push to try and end up with a green try. This shouldn't be reviewer responsibility to verify this, except when reviewing contributor's patches or folks that just joined. More generally, everything that is automated should be handled by the developer, not the reviewer.
  The only thing to verify here is that all code is covered by a test and ask to write one if necessary.
- we are ok maintaining this
  I'm not sure we always ask our self this question. But when it comes to patches adding features, we should try to ensure this new feature is going to be maintained. That someone will be able to fix any regression related to this, any intermittent, bug failure, ...

## Empathetic and inclusive language

I don't feel that was ever an issue, but may be I was lucky enough to not have issues about this.
It looks like a big part of this should be related to the code of conduct.

## Scope of requested changes

## Prefer small PRs, and request to split up large reviews

+1 I think there is room for improvement in our team about this.
I must say that @jryans and @juliandescottes are very good at this already :)

## Be explicit about required and optional changes

I would like to refer to this old tweet from Mossop:
https://twitter.com/EnglishMossop/status/658718135349039104
> Code review is not about obeying the reviewer. It’s a conversation about how code can be better. Sometimes the reviewer is wrong. Push back

I rather see the review as a discussion. All points highlighted by the reviewer should be addressed.
Either by a reply in bugzilla/github or a tweak to the patch. Unless the reviewer explicitly listed it as a "nit", or phrase it as an informational comment only.
I would like to see more discussions during reviews. Hopefully review's comments don't become a checklist of things to fix or ignore.
(This resonates with @jryans comment about considering the work the reviewer did)

## Are there any missing top level items?

See my first paragraph.

## Additional comments

I can easily see this thread ending up with a pretty long guideline.
That will be helpful, as everyone may improve in different particularities of reviews.
But that would be great to identify the main points we should be improving on and followup on them.

Also we have to acknowledge why the review quality dropped. We renewed a big part of the team in a very short timeframe. We missed the opportunity to pass the knowledge around because of that. So thanks for starting this tread, this should help us streamline and spread this skill :)

**Comment by juliandescottes** [comment]
## What is the answer to: "Why do we do code reviews?"

Echoing points mentioned above, we are doing code reviews to improve the quality of the code landing, the consistency of the codebase and spread the knowledge about what is being implemented. 

In previous companies, we sometimes assigned secondary reviewers who wanted to ramp up on a specific part of the codebase. That could also be a goal, although so far in DevTools I have mostly seen one reviewer per commit.

In my opinion we also do reviews to share the responsibility of landing changes. I think the reviewer and the developer are equally responsible for a change that lands in the code base. Knowing that someone else took the time to review your code, to test your change and make sure everything was sane helps me land things with a peaceful mind. But that's one of the definitions I'm not sure is shared by everybody.

## Empathetic and inclusive language

I like the guidelines shared here. I'm usually trying to phrase my comments as suggestions, but it is nice to have this clearly said somewhere. The point about not using "you" is also very interesting.

## Scope of requested changes

Very good point. I loved using code review time to look at potential improvements for follow-ups, but too often my comments were understood as requests. Maybe that's because I phrase most of my comments as suggestions already, so it's hard for the reviewee to know that by "follow up", I really meant follow up! 

What I'm trying to do now is to only comment on what is really relevant for the code review. And for the rest, either file separate bugs or chat about it somewhere else.

## Prefer small PRs, and request to split up large reviews

Totally agree. And reviewers should feel free to reject patches that are too big.

I don't think we need strict guidelines here, as it should be up to the reviewer to decide what is the reasonable patch size for the review. But having good examples of nice patch queues might help. 

## Be explicit about required and optional changes

### How do we actually resolve conflict of when the author and reviewer disagree?

Open an RFC about it? On a more serious note, I hope this should never happen. We should really encourage people to use follow-ups when a review is blocked on a technical issue. If it's a broader issue, then of course involving peers via IRC/Slack/RFC is an option.

### When is it OK to merge something when you disagree with what a reviewer says?

If the comment is optional then it's fine. If it's not optional, you can push back, but the reviewer and reviewee must come to an agreement before landing.

### How much do we trust that the author of the code will do the right thing? Do we have to follow-up with every change before approval?

For new contributors, check everything until landing (also because a sentence such as "r+ w/ green try & comments addressed, thx" will not make any sense to them). 

For regular contributors, I trust them to apply the comments unless said comments go beyond implementation suggestions. For instance:
- found a bug non trivial to fix
- requested architectural changes
- missing tests 

### Is it OK to let small code quality issues slip in?

If it is an issue (not just code style preference), no, it should not slip in. 
If you think it has a negative impact, and you can explain why, then you should comment about it.

Of course the reviewee can disagree, they might have a different perspective on the issue. We all have different backgrounds, we value certain code smells differently. But I think we should still mention them.

It's hard to give a clear yes or no answer here, code quality is a broad term. I think gathering examples could help.

### Is it OK to fail forward, and fix things in follow-ups?

I think we should allow to move things to follow-ups if the current patch is acceptable for the final product (eg Firefox for mozilla-central). Never assume a follow-up will be addressed in the same cycle. Optional comments can naturally transition to follow ups.

@jryans mentioned tests, that's one of the last things I would like to see moved to follow-ups, but I still let it happen from time to time :( The code we ship should be tested.

## Are there any missing top level items?

I don't think so.

## Additional comments

Would it be interesting to review our reviews a bit as we write this? We could monitor our review comments, see what kind of defects we usually log etc... I'm not suggesting to do it on a regular basis, but maybe for a short period to help gather some good examples for this document?

**Comment by sole** [comment]
thanks for this RFC, @gregtatum!

My answers:

## What is the answer to: "Why do we do code reviews?"

As a still-newbie in DevTools code base: I use reviews to ensure I didn't mess up. So I'm using the brains of experienced developers to verify that I'm not doing something I shouldn't be doing. 
Modifying Mozilla code tends to feel really scary so a review takes away a bit of that. I also expect to learn more from you so if there are better ways to do something, I want to hear from you at review time, as the code is still fresh in my mind and I can go back and do the change immediately :-)

## Empathetic and inclusive language

Subscribing to the comments before re: English being a second language for most people here. That said, I'd suggest we strive to be polite and nice. Using 'please' goes a long way in making others feel like colleagues, not subordinates.

We should also ask questions and try to understand where is the author coming from. Maybe they don't know about X, or understand it differently. So that's the 'empathy' we want to develop.

## Scope of requested changes

I think we should feel empowered to ask authors to stay on track and not introduce changes that are irrelevant to the description of the bug (otherwise tracking regressions are harder).

## Prefer small PRs, and request to split up large reviews

Yes

## Be explicit about required and optional changes

Yes to being explicit. You could also be as explicit as r- when there are items that absolutely need addressing.
I like the idea that we follow up with newer contributors and ensure everything is covered with them, but leave some room for experienced staff. However I would try to do that only for smaller issues (what you all mentioned re missing a dot on a comment etc)

## Are there any missing top level items?

Should we require a green try build before performing any serious review? By serious I mean not a review of a 'work in progress' patch where you're seeking early feedback and many things might not be ready for merging. I mean a review for a patch that is supposed to be ready to be landed. If it is ready, it should pass tests. So save some time for the reviewer and run a try build ahead of time, so defects can be found? e.g. linting, telemetry stuff that only happens at build time, etc?

## Additional comments

😀

**Comment by bgrins** [comment]
From https://github.com/devtools-html/rfcs/issues/41#issuecomment-369653517:

> So save some time for the reviewer and run a try build ahead of time, so defects can be found? e.g. linting, telemetry stuff that only happens at build time, etc?

For linting in particular, there's a "Code Review Bot" that started recently running on patches submitted to reviewboard. For example: https://bugzilla.mozilla.org/show_bug.cgi?id=1438482#c12. Which is great since it gives the author quicker feedback and saves time for the reviewer. I think there are similar tools we could integrate with our github repos (if we haven't already).

**Comment by julienw** [comment]
[somewhat OT]
> For linting in particular
Actually I got a linting hook automatically registered on my git cinnabar repository, that's triggered when I run `git mozreview push`. So that's even better. I'm not sure how it got registered, maybe with `mach bootstrap` or `mach mercurial-setup` (I ran the latter to install the `version-control-tools` we need for mozreview, I think).

**Comment by jasonLaster** [comment]
## What is the answer to: "Why do we do code reviews?"
We do reviews for several reasons
1) communicate and build consensus around ideas
2) validate / check for edge cases
3) bike shed on commas. I’m kidding, we shouldn’t do that…

## Empathetic and inclusive language

Yes. Amy captured this best here
http://amyciavolino.com/assets/MindfulCommunicationInCodeReviews.pdf 

I like ending w/ a question like, “what do you think?”. In general, assume that the author has thought about this more than you and you want to learn from them.


## Scope of requested changes
Keep it small. Keep it focused in one dimension (product, architecture, style)… If you combine all three it is difficult to follow.

## Prefer small PRs, and request to split up large reviews

Yes. It should be easy to assess the risk of a change set. The risk goes up with each thing that changes.

It is hard to reason about  a review that changes 2 things because you have to ask three questions? 
- how does the system change after code change 1
- how does the system change after code change 2
- how does the system change after code change 1 + 2
If you change 3 things, it is seven questions: (1), (2), (3), (1,2), (1,3), (2,3), (1,2,3) . If you change 4 things, it’s even more combinations! Yikes!

## Be explicit about required and optional changes

Yes. I think it is tempting to include optional changes like new test coverage and type coverage. These changes, can be lumped in or in many cases could come before or after. 

For instance, if you throw them in at the same time the reviewer might correctly spend their time assessing the risky part and not really have the time to think through the added coverage. If you do a separate PR, then the reviewer has the mental cycles to think about the improved coverage. 

## Are there any missing top level items?

- How long should a review sit? 
- How should you feel about a drawn out pr?
- How easy is it for the reviewer to test the changes?
- How do these questions change when working with a contributor?
- When is it appropriate to re-write a section of the review to show an alternate path?

## Additional comments

Thanks for starting the conversation!

**Comment by captainbrosset** [comment]
## What is the answer to: "Why do we do code reviews?"

- Checking for errors, code quality, test coverage, etc.
- Learning! Both the reviewer and reviewee will be learning through the process. Learn a new part of the code, a new efficient way to write code, ... Lot's of stuff actually, tests, localization, etc. 
- Making the DevTools group own the code, not just the author. We write code as a group, not as individuals, because long-term it's the group that maintains it. Having a review discussion shares the ownership of the code, because authors and reviewers all put a little bit of themselves in it, and the code that results isn't just the result of one person's work.

## Empathetic and inclusive language

Definitely a must. It's very important to put yourself in the shoes of the author when reviewing their code. Also, generally being friendly, supportive, and positive goes a long way.
Again, it's not only the author writing the code. In the end, both parties share the responsibility and knowledge. So being rude or just saying that a thing is wrong without providing another solution is a really bad idea.

Being empathetic here also means double-checking what your review comments. Because you're a reviewer and think you have a better solution doesn't mean that's true. Assume the code author has spent more time thinking about this part of the code than you have (if you're the reviewer) and might actually be right, even if you originally thought something was wrong. Doesn't take long to look up the code and double-check.

Also, being inclusive is highly important. We shouldn't make any assumptions about the person requesting a review, or about the person you're asking a review from, and always provide as much information as required, in a way that is as inclusive as possible.
The bug will live forever online, and many people will read it long after the author and reviewers are done.

## Scope of requested changes

Stay focused on the code changes made on the bug, and don't request other improvements that, even if they seem related, really are tangential.

## Prefer small PRs, and request to split up large reviews

Yes! Small, self-contained changes that can be applied one by one and the code would still run.

## Be explicit about required and optional changes

MozReview has a checkbox that lets you choose whether a comment is an "issue" (needs to be addressed) or not (just a simple comment).
But even without this, it's important to set expectations. Again, be empathetic here, will the code author know what you meant? Does it sound more like a question than a request for change?

## Are there any missing top level items?

- As a reviewer, strive to answer in under a week, if not possible, update the bug to let the requester know when you'll get to the review. Or forward it to someone else who has more time.
- Know when you're out of your comfort zone as a reviewer, and don't hesitate to ask for an extra review to someone else. It's fine, you can't know everything, and the more people participate, the more everybody learns, and the more bugs we avoid.

## Additional comments

**Comment by belen-albeza** [comment]
## Empathetic and inclusive language

One thing that might be interesting to take into account, is that this is a public setting. Even though  team members know each other, and may speak to each other in a different way / more blunt in their 1:1 conversations, these code reviews are public. It's a good chance to set the tone and expectations, and give a good example for external contributors.

**Comment by jimblandy** [comment]
# Why do we do code reviews?

Many people have brought up all my favorite points here, but I have a phrasing I like for one aspect of it:

> Approving a patch means that you are the second person who thinks this change is correct and a good idea.

That does imply that you understand the change in detail, and that you would be a reasonable person to ask about the change if the original author is unavailable.

**Comment by nt1m** [comment]
> And reviewers should feel free to reject patches that are too big.

Splitting up patches is time-consuming to do and can be intimidating for new contributors. If this is something that a contributor should do, so it would be nice to offer some help (something like "Please let me know if you need help splitting the patch.") and potentially followup with them on IRC/Slack. I know some patches in Thunderbird never got reviewed/landed because it was blocked on the contributor having trouble splitting the patch.

> Empathetic and inclusive language

For contributors, it's nice to sometimes be more explicit when asking for a change to be made, especially if the change is non-trivial: why a change should be made ? how this change might potentially interact with the rest of the patch ? ...

**Comment by juliandescottes** [comment]
Flagging as accepted. @gregtatum accepted to summarize the discussion here to extract a documentation from it.

We mentioned in previous discussions about this that it would be great to have a google doc for this so that others can comment on it. @gregtatum can you share the link to the doc here once created? Thanks!

**Comment by juliandescottes** [comment]
Document was opened at: https://docs.google.com/document/d/1HXlPFGXgYn0MXv4OoA2PsSP-3vvvjXxDTGKtNeV0W-8/edit# 

Thanks @gregtatum , closing this one.
